direct.go

```diff
@@ -70,127 +68,91 @@ func (d *direct) loadCandidates(initial map[string]*Provider) {
 		}
 
 		for _, c := range sh {
-			log.Trace("Adding candidate")
 			d.masquerades = append(d.masquerades, &masquerade{Masquerade: *c, ProviderID: key})
 		}
 	}
 }
 
-func (d *direct) providerFor(m *masquerade) *Provider {
-	pid := m.ProviderID
+func (d *direct) providerFor(m MasqueradeInterface) *Provider {
+	pid := m.getProviderID()
 	if pid == "" {
 		pid = d.defaultProviderID
 	}
 	return d.providers[pid]
 }
 
-// Vet vets the specified Masquerade, verifying certificate using the given CertPool
+// Vet vets the specified Masquerade, verifying certificate using the given CertPool.
+// This is used in genconfig.
 func Vet(m *Masquerade, pool *x509.CertPool, testURL string) bool {
-	return vet(m, pool, testURL)
-}
-
-func vet(m *Masquerade, pool *x509.CertPool, testURL string) bool {
-	op := ops.Begin("vet_masquerade")
-	defer op.End()
-	op.Set("masquerade_domain", m.Domain)
-	op.Set("masquerade_ip", m.IpAddress)
-
 	d := &direct{
 		certPool:            pool,
 		maxAllowedCachedAge: defaultMaxAllowedCachedAge,
 		maxCacheSize:        defaultMaxCacheSize,
 	}
-	conn, _, err := d.doDial(m)
+	masq := &masquerade{Masquerade: *m}
+	conn, _, err := d.doDial(masq)
 	if err != nil {
-		op.FailIf(err)
 		return false
 	}
 	defer conn.Close()
-	return postCheck(conn, testURL)
+	return masq.postCheck(conn, testURL)
 }
 
-func (d *direct) vet(numberToVet int) {
-	log.Debugf("Vetting %d initial candidates in series", numberToVet)
-	for i := 0; i < numberToVet; i++ {
-		d.vetOne()
+// findWorkingMasquerades finds working masquerades by vetting them in batches and in
+// parallel. Speed is of the essence here, as without working masquerades, users will
+// be unable to fetch proxy configurations, particularly in the case of a first time
+// user who does not have proxies cached on disk.
+func (d *direct) findWorkingMasquerades() {
+	// vet masquerades in batches
+	const batchSize int = 25
+	var successful atomic.Uint32
+
+	// We loop through all of them until we have 4 successful ones.
+	for i := 0; i < len(d.masquerades) && successful.Load() < 4; i += batchSize {
+		d.vetBatch(i, batchSize, &successful)
 	}
 }
 
-func (d *direct) vetOne() {
-	// We're just testing the ability to connect here, destination site doesn't
-	// really matter
-	log.Debug("Vetting one")
-	unvettedMasquerades := make([]*masquerade, 0, len(d.masquerades))
-	for _, m := range d.masquerades {
-		if m.lastSucceeded().IsZero() {
-			unvettedMasquerades = append(unvettedMasquerades, m)
-		}
+func (d *direct) vetBatch(start, batchSize int, successful *atomic.Uint32) {
+	log.Debugf("Vetting masquerade batch %d-%d", start, start+batchSize)
+	var wg sync.WaitGroup
+	masqueradeSize := len(d.masquerades)
+	for j := start; j < start+batchSize && j < masqueradeSize; j++ {
+		wg.Add(1)
+		go func(m MasqueradeInterface) {
+			defer wg.Done()
+			if d.vetMasquerade(m) {
+				successful.Add(1)
+			}
+		}(d.masquerades[j])
```

Comment:
The above is the most important change along with cleaning up the 10 second timeout thing